security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -279,7 +279,12 @@ func (s *sdsservice) StreamSecrets(stream sds.SecretDiscoveryService_StreamSecre
 			// Reset SDS push time for new SDS push.
 			con.sdsPushTime = time.Time{}
 			con.mutex.Unlock()
-			defer recycleConnection(conID, resourceName)
+			defer func() {
+				recycleConnection(conID, resourceName)
+				// Remove the secret from cache, otherwise refresh job will process this item(if envoy fails to reconnect)
+				// and cause some confusing logs like 'fails to notify because connection isn't found'.
+				s.st.DeleteSecret(conID, resourceName)
```

Comment:
This may be outside the scope of this PR, but why do we even have a secret cache... we are already using client-go Informers which literally has its own cache with the same info. We have just implemented a cache on top of another cache, and now we get all of these bugs as a result. You can already directly call the `informer.Secrets().Get(..)` and it will be even faster than our own secret cache

Reply:
I think we can have a good Control of our cache , this is the interface that we define a lot of functions. Yes your proposal is good and rebuild a wheel is sometimes really of high risk and needs more test protection

Reply:
+1 it makes sense to me to directly rely on k8s client-go cache. Can we add a TODO ?
@JimmyCYJ

Reply:
@williamaronli could you please add a TODO here?

Reply:
Ok i will do also add a unit test and TODO here